java/src/test/java/org/rocksdb/SequenceNumberAllocationTest.java

```diff
@@ -0,0 +1,198 @@
+public class SequenceNumberAllocationTest
+{
+    @Parameter(0)
+    WriteOptions writeOptions;
+
+    @ClassRule
+    public static final RocksNativeLibraryResource ROCKS_NATIVE_LIBRARY_RESOURCE = new RocksNativeLibraryResource();
+
+    @Rule
+    public TemporaryFolder dbFolder = new TemporaryFolder();
+
+    @Test
+    public void testSeqnoFlushAndClose(WriteOptions writeOptions)
+            throws RocksDBException
+    {
+        int count;
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            count = fillDatabase(db, writeOptions, 0, 15);
+            db.flush(new FlushOptions());
+            db.closeE();
+        }
+
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            assertTrue("Database should not be empty", db.getLatestSequenceNumber() != 0);
+            assertEquals("Sequence number changed between closing and reopening the database", count, db.getLatestSequenceNumber());
+        }
+    }
+
+    @Test
+    public void testSeqnoCloseWithoutFlush(WriteOptions writeOptions)
+            throws RocksDBException
+    {
+        int count;
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            count = fillDatabase(db, writeOptions, 0, 30);
+            db.closeE();
+        }
+        try (RocksDB db = RocksDB.open(
+                new Options().setCreateIfMissing(true),
+                dbFolder.getRoot().getAbsolutePath())) {
+            assertTrue("Database should not be empty", db.getLatestSequenceNumber() != 0);
+            assertEquals("Sequence number changed between closing and reopening the database", count, db.getLatestSequenceNumber());
+        }
+    }
+
+    @Test
+    public void testSeqnoThreadInterruptAfterFlush(WriteOptions writeOptions)
+            throws InterruptedException
+    {
+        ExecutorService executor = Executors.newFixedThreadPool(1);
+        AtomicLong sequenceNumberAfterFlush = new AtomicLong(0);
+        AtomicBoolean cancellable = new AtomicBoolean(false);
+        Options options = new Options().setCreateIfMissing(true);
+
+        Future<?> task = executor.submit(() -> {
+            try {
+                RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath());
+                int count = fillDatabase(db, writeOptions, 0, 10);
+                assertEquals("Expected 10 values in the database",10, count);
+                db.flush(new FlushOptions());
+                sequenceNumberAfterFlush.set(db.getLatestSequenceNumber());
+                assertEquals("Sequence number after flush should be 10", 10, sequenceNumberAfterFlush.get());
+                cancellable.set(true);
+                while(true) {
```

Comment:
Do we really want to wait forever here? Seems unlikely. What are we waiting for exactly? We should probably set a reasonable timeout or use a semaphore to signal when this should continue.

Reply:
Opened this as a rough draft before switching context. Thanks for reviewing it! I'll address the comments.